app/src/main/java/nerd/tuxmobil/fahrplan/congress/repositories/LectureListTransformer.kt

```diff
@@ -0,0 +1,105 @@
+package nerd.tuxmobil.fahrplan.congress.repositories
+
+import android.support.v4.util.SparseArrayCompat
+import nerd.tuxmobil.fahrplan.congress.models.Lecture
+
+class LectureListTransformer(private val prioritizedRoomProvider: PrioritizedRoomProvider) {
+
+    /**
+     * Transform data loaded from the database to a data structure that we would like to use in
+     * UI code.
+     * Then transform it to the (legacy) data structure that the UI is currently using.
+     */
+    fun legacyTransformLectureList(dayIndex: Int, lectures: List<Lecture>): LegacyLectureData {
+        val dayRoomData = transformLectureList(dayIndex, lectures)
+
+        // Set Lecture.roomIndex to match the index in DayRoomData.roomDataList
+        dayRoomData.roomDataList.forEachIndexed { index, roomData ->
+            roomData.lectures.forEach { it.roomIndex = index }
+        }
+
+        // Map room title to room index
+        val roomsMap = dayRoomData.roomDataList
+                .mapIndexed { index, roomData -> roomData.roomName to index }
+                .toMap()
+
+        // Map column index to room index
+        // This is simple identity map (0 -> 0, 1 -> 1, etc.)
+        val roomList = SparseArrayCompat<Int>()
+        for (index in 0..dayRoomData.roomDataList.size) {
+            roomList.put(index, index)
+        }
```

Comment:
The `transformLectureList` method preserved as shown in this test:

``` kotlin
@Test
fun `non consecutive rooms indices are preserved`() {
    val lecturesInDatabase = listOf(
            createLecture(lectureId = "L0", room = "Four", roomIndex = 2342),
            createLecture(lectureId = "L1", room = "Two", roomIndex = 1),
            createLecture(lectureId = "L2", room = "One", roomIndex = 0),
            createLecture(lectureId = "L3", room = "Three", roomIndex = 4)
    )

    val dayRoomData = transformer.transformLectureList(dayIndex = 0, lectures = lecturesInDatabase)

    val roomIndices = dayRoomData.roomDataList.flatMap { it.lectures }.map { it.roomIndex }
    assertThat(roomIndices).isEqualTo(listOf(0, 1, 4, 2342))
}
```
Please add the test to your test suite.

The `legacyTransformLectureList` method does not preserve them:

``` kotlin
@Test
fun `non consecutive rooms indices are rewritten`() {
    val lecturesInDatabase = listOf(
            createLecture(lectureId = "L0", room = "Four", roomIndex = 2342),
            createLecture(lectureId = "L1", room = "Two", roomIndex = 1),
            createLecture(lectureId = "L2", room = "One", roomIndex = 0),
            createLecture(lectureId = "L3", room = "Three", roomIndex = 4)
    )

    val lectureData = transformer.legacyTransformLectureList(dayIndex = 0, lectures = lecturesInDatabase)

    val roomIndices = lectureData.lectureList.associate { it.room to it.roomIndex }
    assertThat(roomIndices).isEqualTo(mapOf("One" to 0, "Two" to 1, "Three" to 2, "Four" to 3))
}
```

I am not sure which consequences it has to change the behavior of `FahrplanFragment#loadLectureList`. The `LecturesByRoomIndex` respects non consecutive room indices.

Reply:
> The transformLectureList method preserved [roomIndex] as shown in this test

That's because the values are not touched at all. The new data structure doesn't need this value and we want to get rid of it. So I don't see why you'd want to add a test for that.

> The legacyTransformLectureList method does not preserve them

The legacy UI code uses the roomIndex value to look up the column index. `FahrplanFragment#loadLectureList` tried to make this work with the original roomIndex value but rewrote it if there was a conflict. The new code almost always rewrites the roomIndex value. If some UI code breaks when the roomIndex value in Lecture is not the same as in the database, this PR won't make the situation worse, only more obvious.

Reply:
Thanks for the clarification - once again. Please state the intention of this change in the comments above (line 27). Something like:

``` kotlin
// Map column index to room index
// This is simple identity map (0 -> 0, 1 -> 1, etc.)
// Room indices provided by the database are discarded here on purpose.
// Future UI code which assembles room columns should not be based on these room indices.
```

Please feel free to improve the text.